### DISCUSSION TITLE
Successful TSDB block upload via mimirtool, but no old data returns upon query

### DISCUSSION THREAD

**Discussion by asaf400** [grafana/mimir#3735] [General]
I am not sure where to post this,
It's both a docs issue, and a help & support community request,
sorry if this is spam, here goes:

Hello, after a setup of Kube-Prometheus-Stack + Mimir **_(single tenant)_** on a local dev k8s cluster*,
I am able to see that Prometheus pushes new data into Mimir succesfully, but I would also like to 'import' \ backfill,
the old Prometheus data, which was present locally before remote_write was enabled in the config..

Since I wanted to upload a very small timerange, metrics period of less than 2 hours (prometheus chunk commit interval),
so that prometheus didn't even create a complete tsdb 'chunk',
I have used prometheus snapshot api call, to manually create a tsdb snapshot, 
I then copied the snapshot data directories which looks like
<img width="660" alt="2022-12-15 12_10_17-New Discussion · grafana_mimir" src="https://user-images.githubusercontent.com/14169006/207832201-ccc640c2-79cd-4428-a759-ba345046683c.png">

and used the backfill command to upload:
<img width="1281" alt="2022-12-15 12_12_09-New Discussion · grafana_mimir" src="https://user-images.githubusercontent.com/14169006/207832580-5e401d37-10d1-44e2-b480-e60b1134db7e.png">

But the data doesn't show up in grafana upon query,
So I have 2 explore tabs, one directed at the prometheus instance,
and the other tab directed at the Mimir instance (read endpoint)

with the same simple query:
`container_memory_usage_bytes{pod="application"}`

but the Mimir tab only ever shows the data since the remote_write was enabled,
and nothing from before, while the Prometheus has the complete range of data.

Screenshot of Prometheus results:
<img width="1063" alt="2022-12-14 15_31_02-Explore - Prometheus - Grafana" src="https://user-images.githubusercontent.com/14169006/207835073-66786214-0d7f-4ddd-bf81-d1b9e6efb905.png">

Screenshot of Mimir results (after tsdb snapshot block uploaded, only the fresh data from since remote_write is available) :
<img width="1064" alt="2022-12-14 15_31_30-Explore - Mimir - Grafana" src="https://user-images.githubusercontent.com/14169006/207835131-22103a26-35e4-48bb-9c2e-b31d795df25e.png">

While reading the docs, I got the notion that I might be required to run a flush on the results-cache,
But I do not know how to perform such an action..
Under these docs articles:
https://grafana.com/docs/mimir/latest/operators-guide/configure/configure-tsdb-block-upload/
https://grafana.com/docs/mimir/latest/operators-guide/reference-http-api/
https://grafana.com/docs/mimir/latest/operators-guide/tools/mimirtool/#backfill

Most importantly this quote:
```The results-cache needs flushing
After uploading one or more blocks, the results-cache needs flushing. The reason is that Grafana Mimir caches query results for queries that don’t touch the most recent 10 minutes of data. After uploading blocks however, queries may return different results (because new data was uploaded). Therefore cached results may be wrong, meaning the cache should manually be flushed after uploading blocks.
```

but unfortunately it is not stated in the docs exactly how to achieve this,
it is however stated that it's part of the query mechanism..?
and in the API docs, there is this API: GET,POST /ingester/flush
https://grafana.com/docs/mimir/latest/operators-guide/reference-http-api/#flush-chunks--blocks
Is that related?

** local cluster used in order to facilitate understanding about the migration from pure Prometheus --> Mimir in our production clusters

Edit: multitenant=false

**Comment by pstibrany** [comment]
Hi. There are few possible reasons for this behaviour.

One thing to keep in mind is that backfill uploads blocks to the long term storage. Blocks on long-term storage are queried only after compactor finds them and includes them in "bucket-index", and after that queriers and store-gateways need to reload bucket index. This can take up to 30 minutes in the default configuration. (Config options: `-compactor.cleanup-interval` on compactor and `-blocks-storage.bucket-store.sync-interval` on querier and store-gateway. Both options default to 15 minutes).

Another possibility is that queriers don't actually query blocks from long-term storage if the query falls within most recent 12h (eg. query like "now-3h ... now"). In that case queriers only ask ingesters for data, but the block uploaded via backfill is not stored in ingesters. (Config options: `-querier.query-store-after`, defaults to 12 hours).

To flush the results cache one can simply restart memcachd server, or execute `flush_all` command via memcached port 11211. Both approaches work fine, second one avoids a restart.

**Reply by pstibrany** [reply]
> To flush the results cache one can simply restart memcachd server, or execute flush_all command via memcached port 11211. Both approaches work fine, second one avoids a restart.

There is now a simpler option in coming Mimir 2.8: you can set TTL for cached results for a tenant that is using block upload.

**Comment by asaf400** [comment]
Thanks for the insight, since I start working on this yesterday, I now see the data,
I spun up the cluster today just now, and indeed the old data is showing up in grafana as expected ✅ ,
so that now the earliest data point from prometheus query is also the earliest data point from Mimir..

Thanks for the insights, and specifying the configuration parameters that are related to this topic 👍

**Comment by velu4689** [comment]
@asaf400 : Can you pls share you mimir config ? Also how you integrated mimir with grafana ? I am stuck where remote write isn't working as expected and could not see the data in grafana

**Reply by asaf400** [reply]
I missed this question, I'll try and dig those up, but the config might have been lost, since it was a one time thing..

**Comment by samarara** [comment]
@asaf400 how long did it take you to see the data you backfilled via mimir? I completed a backfill about 3 hours ago and refreshed the results-cache via 
```
nc localhost 11211
flush_all
OK
```
I see the recent data that prometheus is remote writing but I'm unable to see any historical data

**Reply by asaf400** [reply]
I think eventually it was close to the index rollover interval, which by default is 1 day or 12hours, if I remember correctly,
but the data did show up, but I don't think I managed to speed the process by flushing the cache or restarting pods, I just waited..
